lib/Test/Nginx/Socket/Lua/Dgram.pm

```diff
@@ -0,0 +1,445 @@
+my $counter = 0;
+
+add_block_preprocessor(sub {
+    my ($block) = @_;
+
+    my $name = $block->name;
+
+    my $dgram_config = $block->dgram_config;
+    my $dgram_server_config = $block->dgram_server_config;
+    my $dgram_server_config2 = $block->dgram_server_config2;
+    my $dgram_server_config3 = $block->dgram_server_config3;
+    my $dgram_req = $block->dgram_request;
+    my $dgram_req2 = $block->dgram_request2;
+    my $dgram_req3 = $block->dgram_request3;
+
+    if (defined $dgram_server_config || defined $dgram_config) {
+        $dgram_server_config //= '';
+        $dgram_config //= '';
+
+        my $new_main_config = <<_EOC_;
+stream {
+$dgram_config
+    server {
+        listen $port udp;
+
+$dgram_server_config
+    }
+_EOC_
+
+        if (defined $dgram_server_config2) {
+            my $port2 = $port + 1;
+            $new_main_config .= <<_EOC_;
+    server {
+        listen $port2;
+
+$dgram_server_config2
+    }
+_EOC_
+        }
+
+
+        if (defined $dgram_server_config3) {
+            my $port3 = $port + 2;
+            $new_main_config .= <<_EOC_;
+    server {
+        listen $port3;
+
+$dgram_server_config3
+    }
+_EOC_
+        }
+
+        $new_main_config .= "}\n";
+
+        my $main_config = $block->main_config;
+        if (defined $main_config) {
+            $main_config .= $new_main_config;
+        } else {
+            $main_config = $new_main_config;
+        }
+
+        $block->set_value("main_config", $main_config);
+
+        my $new_http_server_config = <<_EOC_;
+            lua_socket_log_errors off;
+
+            location = /t {
+                content_by_lua_block {
+                    local sock, err = ngx.socket.udp()
+                    assert(sock, err)
+
+                    local ok, err = sock:setpeername("$ServerAddr", $port)
+                    if not ok then
+                        ngx.say("connect to stream server error: ", err)
+                        return
+                    end
+_EOC_
+
+        if (defined $dgram_req) {
+            my $file = gen_data_file($dgram_req);
+            $new_http_server_config .= <<_EOC_;
+                    local f = assert(io.open('$file', 'r'))
+                    local data = assert(f:read("*a"))
+                    assert(f:close())
+                    local bytes, err = sock:send(data)
+                    if not bytes then
+                        ngx.say("send stream request error: ", err)
+                        return
+                    end
+_EOC_
+        }
+
+        if (defined $block->abort) {
+            my $timeout = Test::Nginx::Util::parse_time($block->timeout)
+                          // Test::Nginx::Util::timeout();
+            $timeout *= 1000;
+            $new_http_server_config .= <<_EOC_;
+
+                    sock:settimeout($timeout)
+_EOC_
+            $block->set_value("timeout", undef);
+            $block->set_value("abort", undef);
+        }
+
+        $new_http_server_config .= <<_EOC_;
```

Comment:
This part is wrong. Since when `dgram_request` is specified, then the default trigger request is still being sent, leading to two datagrams being sent.